File: 2022_11_03_leetcode_in_fsharp/LeetCode/P0024_SwapNodesInPairs.fs
// url: https://leetcode.com/problems/swap-nodes-in-pairs/
// tags: list, pattern-matching, rec
// examples: [1,2,3,4] -> [2,1,4,3]

module LeetCode.P0024_SwapNodesInPairs

open Utils

let rec swapPairs lst =
    match lst with
    | a :: b :: tail -> b :: a :: swapPairs tail
    | _ -> lst

swapPairs [ 1; 2; 3; 4 ] === [ 2; 1; 4; 3 ]
swapPairs ([]: list<int>) === []
swapPairs [ 1 ] === [ 1 ]

File: 2022_11_19_leetcode_in_csharp/LeetCode/P0024_SwapNodesInPairs.cs
// https://leetcode.com/problems/swap-nodes-in-pairs/

namespace LeetCode;

using static LList;

class P0024_SwapNodesInPairs
{
    static LList<T>? SwapPairs<T>(LList<T>? list)
    {
        return list switch
        {
            (var head1, (var head2, var tail)) => new(head2, new(head1, SwapPairs(tail))),
            _ => list
        };
    }

    public static void Run()
    {
        var inputs = new[]
        {
            LListFrom(1,2,3,4),
            LListFrom(1),

        };

        foreach (var list in inputs)
        {
            var output = SwapPairs(list);
            Console.WriteLine($" {list.ToEnumerable().JoinToString()} -> {output.ToEnumerable().JoinToString()}");
        }
    }
}
